hazelcast/src/main/java/com/hazelcast/internal/dynamicclassloading/DynamicClassloadingService.java

```diff
@@ -0,0 +1,101 @@
+import com.hazelcast.internal.dynamicclassloading.impl.ClassLocator;
+import com.hazelcast.internal.dynamicclassloading.impl.ClassSource;
+import com.hazelcast.internal.dynamicclassloading.impl.filter.Filter;
+import com.hazelcast.internal.dynamicclassloading.impl.ClassDataProvider;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.spi.ManagedService;
+import com.hazelcast.spi.NodeEngine;
+
+import java.util.Properties;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+import static com.hazelcast.internal.dynamicclassloading.impl.filter.ClassNameFilterParser.parseClassNameFilters;
+import static com.hazelcast.internal.dynamicclassloading.impl.filter.MemberProviderFilterParser.parseMemberFilter;
+
+public final class DynamicClassloadingService implements ManagedService {
+    public static final String SERVICE_NAME = "cl-service";
+
+    private boolean enabled;
+
+    private ClassDataProvider provider;
+    private ClassLocator locator;
+
+    @Override
+    public void init(NodeEngine nodeEngine, Properties properties) {
+        DynamicClassloadingConfig config = nodeEngine.getConfig().getDynamicClassloadingConfig();
+        if (!config.isEnabled()) {
+            //TODO: Fail-fast when some attributes are configured but the service is disabled
+            return;
+        }
+
+        ClassLoader parent = nodeEngine.getConfigClassLoader().getParent();
+        Filter<String> classNameFilter = parseClassNameFilters(config);
+        Filter<Member> memberFilter = parseMemberFilter(config.getProviderFilter());
+        ConcurrentMap<String, ClassSource> classMap = new ConcurrentHashMap<String, ClassSource>();
+
+        DynamicClassloadingConfig.ProviderMode providerMode = config.getProviderMode();
+        ILogger providerLogger = nodeEngine.getLogger(ClassDataProvider.class);
+        provider = new ClassDataProvider(providerMode, parent, classMap, providerLogger);
+
+        DynamicClassloadingConfig.ClassCacheMode classCacheMode = config.getClassCacheMode();
+        locator = new ClassLocator(classMap, parent, classNameFilter, memberFilter, classCacheMode, nodeEngine);
+        enabled = config.isEnabled();
+    }
+
+    //called by operations sent by other members
+    public ClassData getClassDataOrNull(String className) {
+        if (!enabled) {
+            return null;
+        }
+        return provider.getClassDataOrNull(className);
+    }
+
+    //called by distributed classloader on this member
+    public Class<?> handleClassNotFoundException(String name)
+            throws ClassNotFoundException {
+        if (!enabled) {
+            throw new ClassNotFoundException("Dynamic Classloading is not enabled.");
+        }
+        return locator.handleClassNotFoundException(name);
+    }
+
+    //called by distributed classloader on this member
+    public Class<?> findLoadedClass(String name) {
+        if (!enabled) {
```

Comment:
Is `enabled` guaranteed to be visible to all threads those are able to load classes?